### ISSUE TITLE
Truncate files when opening them for StandardOutput=file:…

### ISSUE THREAD

**Issue by lucaswerkmeister** [systemd/systemd#8983]
Unlike a standard shell redirection, `StandardOutput=file:…` opens the file without the `O_TRUNC` flag, so that the file pointer is placed at the beginning of the file, but the contents of the file are retained. If the process never does anything except simple writing (no `lseek`, no `ftruncate`, …), and it writes less bytes than the file previously contained, then after the process exits the file will contain a combination of old and new contents:

```
# systemd-run --wait -p StandardOutput=file:/tmp/test echo 123456
# systemd-run --wait -p StandardOutput=file:/tmp/test echo 123
# cat /tmp/test
123
56
```

This does not seem especially useful, so I suggest adding `O_TRUNC` to the code that opens the file. (The patch itself is trivial, but I figured the discussion would be more visible if I filed a separate issue first.)

The documentation currently does not explicitly say that the file is *not* truncated, so I’m not sure how many users already rely on the current behavior, but I suppose this is still a breaking change. WDYT?

Workaround: Truncate the file explicitly, e. g. with `/usr/bin/truncate --size=0 /tmp/test`. (Note that this command will already run with stdout and possibly stderr attached to the file!)

**Comment by lucaswerkmeister** [comment]
Counterpoint:

> If standard input and output are directed to the same file path, it is opened only once, for reading as well as writing and duplicated. This is particular useful when the specified path refers to an `AF_UNIX` socket in the file system, as in that case only a single stream connection is created for both input and output.

In this case, truncating the file is probably not intended… so what do we do here? I haven’t thought this part through yet.

**Comment by keszybz** [comment]
open(2) says
> O_TRUNC
> If the file is a FIFO or terminal  device  file,  the  O_TRUNC flag is ignored. Otherwise, the effect of O_TRUNC is unspecified.

I always assumed it's ignored for pipes under Linux.

**Comment by lucaswerkmeister** [comment]
Sorry, I’m not sure what you mean… doesn’t the manpage specify the effect for pipes (FIFO, ignored, as you assumed)? Or do you mean something else?

(I was worried about sockets for a bit yesterday, but then I learned that you can’t `open` a socket (that results in `ENXIO`, and systemd’s `acquire_path` already uses that).)

**Comment by keszybz** [comment]
It specifies to behaviour for fifos, i.e. "named pipes". I doesn't explicitly say what happens for "normal" pipes. You can create a normal pipe using pipe(2) or pipe2(2) and then reopen it using `open("/dev/fd/3", O_TRUNC|O_WRONLY)`.

I now experimented a bit, and it seems that O_TRUNC is simply ignored in this case, as expected.

**Comment by lucaswerkmeister** [comment]
You mean, if someone specifies `StandardOutput=file:/proc/self/fd/4` in the unit file? I suppose we should prevent that case anyways, to avoid exposing internal file descriptors to the unit.

**Comment by lucaswerkmeister** [comment]
`stat -L /proc/PID/fd/FD` for a pipe FD of some process reports `fifo` type, so perhaps those virtual files really are FIFOs as far as the kernel is concerned?

(And also, `StandardOutput=file:/proc/1/fd/3` results in an error, so it looks like systemd prevents that already.)

**Comment by poettering** [comment]
I am not sure it's safe to generally specify O_TRUNC... Before switching to O_TRUNC by default I figure we should a survey if it really doesn't mean anything on stuff that isn't a regular file, where this option might be used for, i.e. device nodes, fifos, magic stuff in /proc, /sys, /dev/mqueue and such...

Also, I wonder if people might in some cases want O_APPEND instead? hence maybe we should instead introduce trunc: or append: which would be like file: but use the right switch?

**Comment by keszybz** [comment]
open(2) says it's ignored for fifos and "terminal device files". It's also ignored for normal pipes. The other stuff — we need to check.

**Comment by towo** [comment]
I'll chime in saying that I've started using `StandardOutput` for dropping simple metrics result to files. While I can just rm the file in `ExecStartPre`, I'd rather be able to just convince systemd to set `O_TRUNC`; for this case, this can happily be an option in the unit file I'd just specify.

**Comment by bviktor** [comment]
I'd gladly go the ExecStartPre route, but I can't, since those commands are run by the unit user, but [the StandardOutput file is always owned by root:root](https://github.com/systemd/systemd/issues/14385), so if your unit is run as anything else but root, you're basically out of options.

**Comment by bviktor** [comment]
Actually, come to think of it, you _can_ rm the file (but not truncate it). So that means now the handle changes everytime, so any `tail`s need to be restarted. Geez...

In any case, I'd be more than interested to hear the use case for starting to overwrite an existing file from the beginning. I can't come up with any.

**Comment by lucaswerkmeister** [comment]
> I'd gladly go the ExecStartPre route, but I can't, since those commands are run by the unit user, but [the StandardOutput file is always owned by root:root](https://github.com/systemd/systemd/issues/14385), so if your unit is run as anything else but root, you're basically out of options.

Not to go too off-topic, but I would assume that `ExecStartPre=+/usr/bin/...` should work (the `+` prefix runs the executable with full privileges, ignoring `User=` etc.).

**Comment by bviktor** [comment]
TIL. That's slightly better, thanks!

**Comment by poettering** [comment]
Happy to review/merge a patch that adds `truncate:`, similar in style to the `append:` we already have.

**Comment by bviktor** [comment]
Awesome, thanks a lot!
